### PR TITLE
Feature/python packaging

### PR DESCRIPTION
**Pull request by migwellian**
This PR makes it possible for us to deploy our python package to PyPi so that users can `pip install keanu`.

Currently you also need to put all your ND4J dependencies into a local directory and point to it with an environment variable `KEANU_ND4J_CLASSPATH`. This is only because of its size (around 250MB) - there is a 60MB limit on PyPi, although this will be extended if we request it. 

Next step: see if we can strip ND4J down to just the bits we need, and measure its size.

### CODE REVIEW

File: keanu-python/setup.py [hunk 1]
```diff
@@ -7,12 +7,13 @@ def readme():
 setup(name='keanu',
       description='A probabilistic approach from an Improbabilistic company',
       long_description=readme(),
-      version='0.0.14.dev0',
+      version='0.0.14.dev8',
```
Comment (icantsayrural) [inline keanu-python/setup.py]: `dev0`?

File: keanu-python/setup.py [hunk 2]
```diff
@@ -7,12 +7,13 @@ def readme():
...
       author='Improbable Worlds',
       author_email='keanu-engineering@improbable.io',
       url='https://github.com/improbable-research/keanu',
       license='MIT',
-      packages=['keanu'],
+      packages=find_packages(),
```
Comment (icantsayrural) [inline keanu-python/setup.py]: 🎉

File: keanu-python/keanu/base.py [hunk 3]
```diff
@@ -1,12 +1,23 @@
...
+            ) from e        
         self._class = self.unwrap().getClass().getSimpleName()
 
     def __repr__(self):
-        args = [str(arg) for arg in self._args]
-        return "[{0} => {1}: ({2})]".format(self._class, type(self), ",".join(args))
+        return "[{0} => {1}: {2}]".format(self._class, type(self), JavaObjectWrapper.print_signature(self._ctor, self._args))
+
+    @staticmethod
+    def print_signature(_ctor, _args):
```
Comment (icantsayrural) [inline keanu-python/keanu/base.py]: Make this private?

**Comment by migwellian** [comment]
Here is a list of the largest jar files amongst ND4J's dependencies:

```
-rw-r--r--  1 michaellewin  staff   668K 29 Oct 16:26 openblas-0.3.0-1.4.2-ios-arm64.jar
-rw-r--r--  1 michaellewin  staff   777K 29 Oct 16:26 openblas-0.3.0-1.4.2-android-arm.jar
-rw-r--r--  1 michaellewin  staff   859K 29 Oct 16:26 openblas-0.3.0-1.4.2-android-arm64.jar
-rw-r--r--  1 michaellewin  staff   871K 29 Oct 16:26 openblas-0.3.0-1.4.2-ios-x86_64.jar
-rw-r--r--  1 michaellewin  staff   1.2M 29 Oct 16:26 openblas-0.3.0-1.4.2-android-x86.jar
-rw-r--r--  1 michaellewin  staff   1.2M 29 Oct 16:26 nd4j-native-1.0.0-beta2.jar
-rw-r--r--  1 michaellewin  staff   1.4M 29 Oct 16:26 protobuf-java-shaded-351-0.9.jar
-rw-r--r--  1 michaellewin  staff   1.4M 29 Oct 16:26 openblas-0.3.0-1.4.2-android-x86_64.jar
-rw-r--r--  1 michaellewin  staff   1.5M 29 Oct 16:26 mkl-2018.3-1.4.2-windows-x86.jar
-rw-r--r--  1 michaellewin  staff   1.8M 29 Oct 16:26 jackson-1.0.0-beta2.jar
-rw-r--r--  1 michaellewin  staff   1.9M 29 Oct 16:26 commons-math3-3.5.jar
-rw-r--r--  1 michaellewin  staff   2.3M 29 Oct 16:26 mkl-2018.3-1.4.2-windows-x86_64.jar
-rw-r--r--  1 michaellewin  staff   2.3M 29 Oct 16:26 guava-20.0.jar
-rw-r--r--  1 michaellewin  staff   2.5M 29 Oct 16:26 mkl-2018.3-1.4.2-macosx-x86_64.jar
-rw-r--r--  1 michaellewin  staff   2.9M 29 Oct 16:26 mkl-2018.3-1.4.2-linux-x86_64.jar
-rw-r--r--  1 michaellewin  staff   3.1M 29 Oct 16:26 mkl-2018.3-1.4.2-linux-x86.jar
-rw-r--r--  1 michaellewin  staff   5.1M 29 Oct 16:26 nd4j-api-1.0.0-beta2.jar
-rw-r--r--  1 michaellewin  staff   5.4M 29 Oct 16:26 nd4j-native-1.0.0-beta2-android-arm.jar
-rw-r--r--  1 michaellewin  staff   5.5M 29 Oct 16:26 nd4j-native-1.0.0-beta2-ios-arm64.jar
-rw-r--r--  1 michaellewin  staff   5.6M 29 Oct 16:26 openblas-0.3.0-1.4.2-linux-armhf.jar
-rw-r--r--  1 michaellewin  staff   5.7M 29 Oct 16:26 nd4j-native-1.0.0-beta2-android-arm64.jar
-rw-r--r--  1 michaellewin  staff   5.8M 29 Oct 16:26 nd4j-native-1.0.0-beta2-linux-ppc64le.jar
-rw-r--r--  1 michaellewin  staff   6.1M 29 Oct 16:26 nd4j-native-1.0.0-beta2-linux-x86_64.jar
-rw-r--r--  1 michaellewin  staff   6.2M 29 Oct 16:26 nd4j-native-1.0.0-beta2-android-x86_64.jar
-rw-r--r--  1 michaellewin  staff   6.4M 29 Oct 16:26 nd4j-native-1.0.0-beta2-ios-x86_64.jar
-rw-r--r--  1 michaellewin  staff   6.5M 29 Oct 16:26 nd4j-native-1.0.0-beta2-macosx-x86_64.jar
-rw-r--r--  1 michaellewin  staff   6.6M 29 Oct 16:26 nd4j-native-1.0.0-beta2-android-x86.jar
-rw-r--r--  1 michaellewin  staff   6.8M 29 Oct 16:26 nd4j-native-1.0.0-beta2-windows-x86_64.jar
-rw-r--r--  1 michaellewin  staff   6.9M 29 Oct 16:26 openblas-0.3.0-1.4.2-linux-ppc64le.jar
-rw-r--r--  1 michaellewin  staff   8.7M 29 Oct 16:26 openblas-0.3.0-1.4.2-linux-x86.jar
-rw-r--r--  1 michaellewin  staff   8.8M 29 Oct 16:26 openblas-0.3.0-1.4.2-windows-x86.jar
-rw-r--r--  1 michaellewin  staff    12M 29 Oct 16:26 openblas-0.3.0-1.4.2-linux-x86_64.jar
-rw-r--r--  1 michaellewin  staff    13M 29 Oct 16:26 openblas-0.3.0-1.4.2-windows-x86_64.jar
-rw-r--r--  1 michaellewin  staff    13M 29 Oct 16:26 openblas-0.3.0-1.4.2-macosx-x86_64.jar
-rw-r--r--  1 michaellewin  staff    31M 29 Oct 16:26 mkl-dnn-0.15-1.4.2-macosx-x86_64.jar
-rw-r--r--  1 michaellewin  staff    41M 29 Oct 16:26 mkl-dnn-0.15-1.4.2-linux-x86_64.jar
-rw-r--r--  1 michaellewin  staff    41M 29 Oct 16:26 mkl-dnn-0.15-1.4.2-windows-x86_64.jar
```

**Comment by migwellian** [comment]
I guess we can exclude `mkl-dnn` altogether...?

File: keanu-python/setup.py (hunk 1 above)
Comment (migwellian) [reply keanu-python/setup.py]: I'm just incrementing this every time I upload a version to pypi - because it fails if the version already exists.

File: keanu-python/keanu/base.py (hunk 3 above)
Comment (migwellian) [reply keanu-python/keanu/base.py]: I think I'm gonna need to nuke this anyway when I merge your PR in

**Comment by migwellian** [comment]
I've got ND4J's dependencies down to 75MB but I don't think I can go any further. We could split our python package up into 3 different versions (for 3 platforms) but that feels like introducing unnecessary complexity.

    2.1K 30 Oct 14:04 mkl-platform-2018.3-1.4.2.jar
    2.1K 30 Oct 14:04 mkl-dnn-platform-0.15-1.4.2.jar
    2.9K 30 Oct 14:04 nd4j-native-platform-1.0.0-beta2.jar
    8.6K 30 Oct 14:04 nd4j-context-1.0.0-beta2.jar
    9.9K 30 Oct 14:04 flatbuffers-1.2.0-3f79e055.jar
     26K 30 Oct 14:04 cloning-1.9.3.jar
     40K 30 Oct 14:04 slf4j-api-1.7.21.jar
     47K 30 Oct 14:04 nd4j-native-api-1.0.0-beta2.jar
     52K 30 Oct 14:04 neoitertools-1.0.0.jar
     54K 30 Oct 14:04 objenesis-2.6.jar
     64K 30 Oct 14:04 nd4j-buffer-1.0.0-beta2.jar
     68K 30 Oct 14:04 protobuf-java-util-shaded-351-0.9.jar
    125K 30 Oct 14:04 openblas-0.3.0-1.4.2.jar
    147K 30 Oct 14:04 mkl-dnn-0.15-1.4.2.jar
    195K 30 Oct 14:04 mkl-2018.3-1.4.2.jar
    204K 30 Oct 14:04 commons-io-2.5.jar
    213K 30 Oct 14:04 nd4j-common-1.0.0-beta2.jar
    227K 30 Oct 14:04 gson-2.7.jar
    264K 30 Oct 14:04 snakeyaml-1.12.jar
    278K 30 Oct 14:04 commons-codec-1.10.jar
    352K 30 Oct 14:04 javacpp-1.4.2.jar
    483K 30 Oct 14:04 commons-lang3-3.6.jar
    546K 30 Oct 14:04 commons-compress-1.16.1.jar
    560K 30 Oct 14:04 joda-time-2.2.jar
    1.2M 30 Oct 14:04 nd4j-native-1.0.0-beta2.jar
    1.4M 30 Oct 14:04 protobuf-java-shaded-351-0.9.jar
    1.8M 30 Oct 14:04 jackson-1.0.0-beta2.jar
    1.9M 30 Oct 14:04 commons-math3-3.5.jar
    2.3M 30 Oct 14:04 guava-20.0.jar
    5.1M 30 Oct 14:04 nd4j-api-1.0.0-beta2.jar
    6.1M 30 Oct 14:04 nd4j-native-1.0.0-beta2-linux-x86_64.jar
    6.5M 30 Oct 14:04 nd4j-native-1.0.0-beta2-macosx-x86_64.jar
    6.8M 30 Oct 14:04 nd4j-native-1.0.0-beta2-windows-x86_64.jar
     12M 30 Oct 14:04 openblas-0.3.0-1.4.2-linux-x86_64.jar
     13M 30 Oct 14:04 openblas-0.3.0-1.4.2-windows-x86_64.jar
     13M 30 Oct 14:04 openblas-0.3.0-1.4.2-macosx-x86_64.jar

**Comment by icantsayrural** [comment]
Is it possible for `KeanuContext` to download Java Keanu from Github and create a jar file?

**Comment by migwellian** [comment]
Request to increase PiPI file size limit: https://github.com/pypa/warehouse/issues/4979

**Comment by migwellian** [comment]
> Is it possible for KeanuContext to download Java Keanu from Github and create a jar file?

We could do something like that, but I don't want to be dependent on internet connectivity at runtime. I think any pulling down of dependencies ought to happen at installation time.

**Comment by migwellian** [comment]
We've been given an increased limit of 100MB now

File: keanu-python/setup.py [hunk 4]
```diff
@@ -1,19 +1,25 @@
 from setuptools import setup, find_packages
-
+import shutil
 
 def readme():
     with open('README.rst', "r") as f:
         return f.read()
 
+version_string = '0.0.14.dev8'
```
Comment (icantsayrural) [inline keanu-python/setup.py]: It could be `0.0.15.dev0` if we want to follow java Keanu.

File: .gitignore [hunk 5]
```diff
@@ -6,4 +6,6 @@ nbproject
 *.iml
 build
 /out
-**/*.gpg
\ No newline at end of file
+**/*.gpg
+keanu-python/dist/
```
Comment (icantsayrural) [inline .gitignore]: `keanu-python` has its own `.gitignore` file

File: keanu-python/keanu/context.py [hunk 6]
```diff
@@ -31,6 +35,14 @@ def __init__(self):
 
         self.__jvm_view = self._gateway.new_jvm_view()
 
+    def __build_classpath(self):
+        keanu_path = os.path.join(PATH, "classpath", "*")
+        nd4j_path = os.environ.get(ND4J_CLASSPATH_ENVIRONMENT_VARIABLE)
+        if nd4j_path is None:
+            return keanu_path
+        else:
+            return os.pathsep.join([keanu_path, os.path.join(nd4j_path, "*")])
```
Comment (icantsayrural) [inline keanu-python/keanu/context.py]: Is this for appveyor?

File: keanu-project/src/test/resources/log4j.properties [hunk 7]
```diff
@@ -0,0 +1,10 @@
+# Root logger option
```
Comment (icantsayrural) [inline keanu-project/src/test/resources/log4j.properties]: Just curious, is this related to this PR?

File: keanu-python/setup.py [hunk 8]
```diff
@@ -1,18 +1,26 @@
...
       description='A probabilistic approach from an Improbabilistic company',
       long_description=readme(),
-      version='0.0.14.dev0',
+      version=version_string,
       author='Improbable Worlds',
       author_email='keanu-engineering@improbable.io',
       url='https://github.com/improbable-research/keanu',
       license='MIT',
-      packages=['keanu'],
+      packages=find_packages(exclude=["examples"]),
```
Comment (icantsayrural) [inline keanu-python/setup.py]: Should we also exclude tests?

File: keanu-python/setup.py [hunk 9]
```diff
@@ -1,18 +1,26 @@
...
+import shutil
 
 def readme():
     with open('README.rst', "r") as f:
         return f.read()
 
+version_string = '0.0.15.dev0'
+
+# If you don't remove the old directories, it tends to put the excluded module "examples" into the bdist
+for dir_name in ("keanu-%s.dist-info" % version_string, "keanu.egg-info", "build", "dist"):
```
Comment (icantsayrural) [inline keanu-python/setup.py]: It seems like we can get around this bug my adding some more stuff to our `MANIFEST.in`

[Source](https://stackoverflow.com/questions/26545668/setup-py-sdist-exclude-packages-in-subdirectory)

File: keanu-python/setup.py (hunk 8 above)
Comment (migwellian) [reply keanu-python/setup.py]: that seems to happen automatically

File: keanu-project/src/test/resources/log4j.properties (hunk 7 above)
Comment (migwellian) [reply keanu-project/src/test/resources/log4j.properties]: it's related to me trying to debug why it was freezing on Windows!

Comment (migwellian) [reply keanu-project/src/test/resources/log4j.properties]: and it will be useful to have this in our main branch

File: keanu-python/keanu/context.py (hunk 6 above)
Comment (migwellian) [reply keanu-python/keanu/context.py]: yes, it's for Windows.

For now, Windows users have to download a 40MB jar (mkl-dnn) and set an environment variable to point to it

Comment (migwellian) [reply keanu-python/keanu/context.py]: 😢

File: .gitignore (hunk 5 above)
Comment (migwellian) [reply .gitignore]: good call - I'll change it

File: keanu-python/setup.py (hunk 9 above)
Comment (migwellian) [reply keanu-python/setup.py]: I don't think so - see e.g. https://github.com/pypa/setuptools/issues/511
